### PR TITLE
Try to delete the workspace instead of fail

### PR DESCRIPTION
**Pull request by jessehouwing**
Creation of the new workspace will fail anyway if it still exists.
But 2 agents may delete the same workspace at the exact same time.

Fixes: #3588

### CODE REVIEW

File: src/Agent.Plugins/TfsVCSourceProvider.cs [hunk 1]
```diff
@@ -253,7 +253,8 @@ public sealed class TfsVCSourceProvider : ISourceProvider
                 tfWorkspaces = await tf.WorkspacesAsync(matchWorkspaceNameOnAnyComputer: true);
                 foreach (ITfsVCWorkspace tfWorkspace in tfWorkspaces ?? new ITfsVCWorkspace[0])
                 {
-                    await tf.WorkspaceDeleteAsync(tfWorkspace);
+
+                    await tf.TryWorkspaceDeleteAsync(tfWorkspace);
```
Comment (alexander-smolyakov) [inline src/Agent.Plugins/TfsVCSourceProvider.cs:257]: Could you please fix the code formatting?
```suggestion
                    await tf.TryWorkspaceDeleteAsync(tfWorkspace);
```

File: src/Agent.Plugins/TfsVCSourceProvider.cs [hunk 2]
```diff
@@ -253,7 +253,8 @@ public sealed class TfsVCSourceProvider : ISourceProvider
                 tfWorkspaces = await tf.WorkspacesAsync(matchWorkspaceNameOnAnyComputer: true);
                 foreach (ITfsVCWorkspace tfWorkspace in tfWorkspaces ?? new ITfsVCWorkspace[0])
                 {
-                    await tf.WorkspaceDeleteAsync(tfWorkspace);
```
Comment (anatolybolshakov) [inline src/Agent.Plugins/TfsVCSourceProvider.cs:256]: Please add message trace for the case when tf.WorkspaceDeleteAsync returns false.

Comment (anatolybolshakov) [reply src/Agent.Plugins/TfsVCSourceProvider.cs:256]: Nvm, the message is already here in WorkspaceDeleteAsync

**Comment by anatolybolshakov** [comment]
@MaximMalygin with these changes agent will skip tfvc workspace removal - if there is an error during it.
Do you have any objections? Could there be any other errors which we can't ignore during workspace removal?

**Comment by jessehouwing** [comment]
The current error I see a lot is that someone else has deleted the workspace between checking for its existence and the subsequent deletion.

If I'd had to guess there is some job that runs when an agent is taken offline in the hosted pool which deletes she workspace from the server.

It such a process exists (can't find any mention for it in the agent itself), then is should be sure no current job has already spawned on the same workspace name. I'm seeing a number of deleted workspaces I can't explain on the hosted pool.

**Comment by jessehouwing** [comment]
In that case, there is no need to fail. The tool will return: 
```
TF14061: The workspace ws_1_326;Project Collection Build Service (OrgS) does not exist.
```

I've though about adding a 2nd query workspaces to be sure after the loop. But I can't see why the delete would fail in another case.

**Comment by RuslanRasulovAkv** [comment]
@anatolybolshakov 
The root problem was that Hosted Agent on the machine name, could remove another Agent's workspace if they both use the same folder ( [https://github.com/jessehouwing/azure-pipelines-agent/blob/9077f4b69892abaf06a3d0ea4d6823a296a6683e/src/Agent.Plugins/TfsVCSourceProvider.cs#L244](url) ). As far as  here ( [https://github.com/microsoft/azure-pipelines-agent/pull/3586](PR 3586) ) we ensured the uniqueness of the folder it should fix that problem.
As for these changes seems they wouldn't broke anything.
cc @MaximMalygin

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!

**Review by alexander-smolyakov — APPROVED** [review]

**Comment by jessehouwing** [comment]
An explicit folder mapping will still break things I'm guessing...

```
$/Project/Something/That/Must/Be/In/A/Fixed/Location => c:\myspeciallocation\
```

**Pull request merged**